File: androidApp/app/src/main/java/com/harnest/app/engine/FsBridge.kt
package com.harnest.app.engine

import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Sync fs bridge — mirrors qjs_fs_call in harness_engine.cpp.
 * Sandbox: all paths resolve under engine.sandboxRoot; prefix-checked.
 */
object FsBridge {

    fun handle(engine: HarnessEngine, reqJson: String): String {
        return try {
            val req = JSONObject(reqJson)
            val op = req.optString("op", "")
            val path = req.optString("path", "")
            val resolved = resolve(engine, path)
                ?: return err("path escapes sandbox: $path")
            when (op) {
                "exists" -> JSONObject().put("exists", resolved.exists())
                "readFile" -> run {
                    if (!resolved.isFile) return@run err("open failed: $path")
                    val bytes = resolved.readBytes()
                    val isText = req.optBoolean("text", true)
                    if (isText) {
                        JSONObject().put("ok", true).put("data", String(bytes, Charsets.UTF_8))
                    } else {
                        JSONObject().put("ok", true).put("data", android.util.Base64.encodeToString(bytes, android.util.Base64.NO_WRAP))
                    }
                }
                "writeFile" -> run {
                    val data = req.optString("data", "")
                    val isB64 = req.optBoolean("base64", false)
                    resolved.parentFile?.mkdirs()
                    try {
                        if (isB64) {
                            resolved.writeBytes(android.util.Base64.decode(data, android.util.Base64.NO_WRAP))
                        } else {
                            resolved.writeText(data, Charsets.UTF_8)
                        }
                        JSONObject().put("ok", true)
                    } catch (e: Throwable) {
                        err("write failed: ${e.message}")
                    }
                }
                "readdir" -> run {
                    if (!resolved.isDirectory) return@run err("not a directory")
                    val entries = JSONArray()
                    resolved.listFiles()?.forEach { entries.put(it.name) }
                    JSONObject().put("ok", true).put("entries", entries)
                }
                "mkdir" -> run {
                    val rec = req.optBoolean("recursive", true)
                    val ok = if (rec) resolved.mkdirs() else resolved.mkdir()
                    JSONObject().put("ok", ok || resolved.isDirectory)
                }
                "rm" -> run {
                    val rec = req.optBoolean("recursive", false)
                    val ok = if (rec) resolved.deleteRecursively() else resolved.delete()
                    JSONObject().put("ok", ok)
                }
                "stat" -> run {
                    if (!resolved.exists()) return@run err("not found: $path")
                    val isDir = resolved.isDirectory
                    JSONObject().put("ok", true)
                        .put("isDir", isDir)
                        .put("size", if (isDir) 0L else resolved.length())
                        .put("mtimeMs", resolved.lastModified())
                }
                "rename" -> run {
                    val newPath = req.optString("newPath", "")
                    val target = resolve(engine, newPath)
                        ?: return@run err("path escapes sandbox: $newPath")
                    if (!resolved.exists()) return@run err("not found: $path")
                    try {
                        target.parentFile?.mkdirs()
                        if (target.exists() && !target.delete()) return@run err("rename failed: target busy")
                        if (resolved.renameTo(target)) JSONObject().put("ok", true)
                        else err("rename failed: $path")
                    } catch (e: Throwable) {
                        err("rename failed: ${e.message}")
                    }
                }
                "realpath" -> JSONObject().put("ok", true).put("path", resolved.path)
                else -> err("unknown op: $op")
            }.toString()
        } catch (e: Throwable) {
            err("fs bridge error: ${e.message}")
        }
    }

    /** Resolve path inside sandbox; null when it escapes. */
    fun resolve(engine: HarnessEngine, path: String): File? {
        val root = engine.sandboxRoot
        return try {
            val f = if (path.startsWith("/")) File(path) else File(root, path)
            val canonical = f.canonicalFile
            val rootCanonical = root.canonicalFile
            if (canonical == rootCanonical || canonical.path.startsWith(rootCanonical.path + File.separator)) {
                canonical
            } else null
        } catch (e: Throwable) {
            null
        }
    }

    private fun err(msg: String): String = JSONObject().put("ok", false).put("error", msg).toString()
}

File: iosApp/Sources/Service/FsBridge.swift
import Foundation

/// Sandboxed fs ops for harness.js — mirrors FsBridge.kt / fs_bridge.cpp.
/// All paths resolve inside the engine sandbox; canonicalized prefix-checked.
enum FsBridge {

    static func handle(engine: HarnessEngine, reqJson: String) -> String {
        guard let req = (try? JSONSerialization.jsonObject(with: Data(reqJson.utf8))) as? [String: Any] else {
            return err("bad fs request json")
        }
        let op = req["op"] as? String ?? ""
        let path = req["path"] as? String ?? ""
        guard let resolved = resolve(engine: engine, path: path) else {
            return err("path escapes sandbox: \(path)")
        }
        let fm = FileManager.default
        var isDir: ObjCBool = false

        switch op {
        case "exists":
            let ok = fm.fileExists(atPath: resolved.path, isDirectory: &isDir)
            return json(["ok": true, "exists": ok, "isDir": isDir.boolValue])

        case "readFile":
            guard fm.fileExists(atPath: resolved.path, isDirectory: &isDir), !isDir.boolValue,
                  let bytes = fm.contents(atPath: resolved.path) else {
                return err("open failed: \(path)")
            }
            if req["text"] as? Bool ?? true {
                return json(["ok": true, "data": String(data: bytes, encoding: .utf8) ?? ""])
            }
            return json(["ok": true, "data": bytes.base64EncodedString()])

        case "writeFile":
            let data = req["data"] as? String ?? ""
            let payload: Data
            if req["base64"] as? Bool ?? false {
                payload = Data(base64Encoded: data) ?? Data()
            } else {
                payload = Data(data.utf8)
            }
            if !resolved.deletingLastPathComponent().path.isEmpty {
                try? fm.createDirectory(at: resolved.deletingLastPathComponent(), withIntermediateDirectories: true)
            }
            if fm.createFile(atPath: resolved.path, contents: payload) {
                return json(["ok": true])
            }
            return err("write failed: \(path)")

        case "readdir":
            guard fm.fileExists(atPath: resolved.path, isDirectory: &isDir), isDir.boolValue,
                  let names = try? fm.contentsOfDirectory(atPath: resolved.path) else {
                return err("not a directory: \(path)")
            }
            return json(["ok": true, "entries": names])

        case "mkdir":
            let recursive = req["recursive"] as? Bool ?? true
            do {
                try fm.createDirectory(atPath: resolved.path, withIntermediateDirectories: recursive)
                return json(["ok": true])
            } catch {
                if fm.fileExists(atPath: resolved.path, isDirectory: &isDir), isDir.boolValue {
                    return json(["ok": true])
                }
                return err("mkdir failed: \(error.localizedDescription)")
            }

        case "rm":
            let recursive = req["recursive"] as? Bool ?? false
            guard fm.fileExists(atPath: resolved.path, isDirectory: &isDir) else {
                return json(["ok": false])
            }
            if isDir.boolValue && !recursive {
                let entries = (try? fm.contentsOfDirectory(atPath: resolved.path)) ?? []
                if !entries.isEmpty { return json(["ok": false]) }
            }
            do {
                try fm.removeItem(atPath: resolved.path)
                return json(["ok": true])
            } catch {
                return json(["ok": false])
            }

        case "rename":
            let newPath = req["newPath"] as? String ?? ""
            guard let target = resolve(engine: engine, path: newPath) else {
                return err("path escapes sandbox: \(newPath)")
            }
            guard fm.fileExists(atPath: resolved.path) else {
                return err("not found: \(path)")
            }
            try? fm.createDirectory(at: target.deletingLastPathComponent(), withIntermediateDirectories: true)
            if fm.fileExists(atPath: target.path) {
                do { try fm.removeItem(atPath: target.path) } catch {
                    return err("rename failed: target busy")
                }
            }
            do {
                try fm.moveItem(atPath: resolved.path, toPath: target.path)
                return json(["ok": true])
            } catch {
                return err("rename failed: \(error.localizedDescription)")
            }

        case "realpath":
            return json(["ok": true, "path": resolved.path])

        case "stat":
            guard fm.fileExists(atPath: resolved.path, isDirectory: &isDir) else {
                return err("not found: \(path)")
            }
            let attrs = try? fm.attributesOfItem(atPath: resolved.path)
            let size = (attrs?[.size] as? NSNumber)?.int64Value ?? 0
            let mtime = (attrs?[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0
            return json([
                "ok": true,
                "isDir": isDir.boolValue,
                "size": isDir.boolValue ? 0 : size,
                "mtimeMs": Int64(mtime * 1000),
            ])

        default:
            return err("unknown fs op: \(op)")
        }
    }

    static func resolve(engine: HarnessEngine, path: String) -> URL? {
        let root = engine.sandboxRoot.standardizedFileURL.resolvingSymlinksInPath()
        let f = path.hasPrefix("/")
            ? URL(fileURLWithPath: path)
            : root.appendingPathComponent(path)
        let canonical = f.standardizedFileURL.resolvingSymlinksInPath()
        let cp = canonical.path
        let rp = root.path
        if cp == rp || cp.hasPrefix(rp + "/") { return canonical }
        return nil
    }

    private static func json(_ obj: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return "{\"ok\":false}" }
        return String(data: data, encoding: .utf8) ?? "{\"ok\":false}"
    }

    private static func err(_ msg: String) -> String {
        json(["ok": false, "error": msg])
    }
}
